[ADD] Open, Close 마일스톤 보여주기 & 마일스톤 Open/Close 상태 변경

marulloc opened the pull request:
[comment]: <> (PR 태그를 명시해주세요. [ADD] / [UPDATE] / [BUG])

### 요약
마일스톤 페이지에서, open/close된 마일스톤끼리 볼 수 있는 기능 추가
마일스톤의 open close 상태를 변경 할 수 있음

### 핵심 로직 및 내용

<br>

**1. opened / closed 마일스톤끼리 보여주기**

- 마일스톤 페이지에서, whichMile 이라는 state를 추가. 
- open버튼을 누르면, whichMile = 1  // close버튼을 누르면 whichMile = 0 으로 state 변경
- 마일스톤 렌더링 할 때, `filter().map()`을 이용 (whichMile과 같은 isOpen 값을 가진 마일스톤만 렌더)

<br>
<br>

**2. 마일스톤 open / close 상태 변경**

###### 프론트 부분
- open된 마일스톤은 close 버튼을 가지고있고
- close된 마일스톤은 open 버튼을 가지고 있다.
- 버튼을 누르면 axios.patch 동작. 

###### 백엔드 부분
- 구조분해 할당으로, params 파싱하여 sql문 생성
- 현재 edit 기능만 동작하지만, 백엔드 부분에서 parsing하는 로직은 이미 작성했슴다.



### 기타 참고 사항
백엔드 `req.params` 쓸 때 
`/milestone/:5` 를 파싱하면  `{ milestoneId : ':5' }` 처럼 `:`가 함께 가져와집니당 ㅠ

marulloc committed:
[FEAT] New MileStone Page

[why]
create new Milestonr

[what]
add new milestone page
add link to NewMilestonePage
add submit button ( disabled if title is empty)
add axios.post code with Credential true
add onChange method at each input tag

[etc]
etc...

marulloc committed:
[FEAT] New Milestone API

[why]
post new Milestone

[what]
add route logic
add db post logic

[etc]
etc...

marulloc committed:
[BUG] Query default value

[why]
bug cause date's default value was ''
when post data into db occur error

[what]
alter default value '' to null

[etc]
etc...

marulloc committed:
[FEAT] Show Open Close Milestone btn

[why]
show milestone separatly based on whether opened or not...

[what]
add state whichMile
click open btn -> whichMile = 1
click close btn -> whichMile = 0

add filter().map()
show milestone which have same value with whichMile's value

[etc]
etc...

marulloc committed:
[FEAT] Milestone Open Close

[why]
for changing Milstone's state

[what]
add Open, Close Btn
add axios.patch logic
add route logic that parsing params & making query statement

[etc]
etc...

marulloc committed:
[STYLE] PARAMS

[why]
send params
get params

[what]
change url in front

[etc]
etc...

thesulks commented on Backend/Routes/milestoneRoute.js:
```diff
@@ -1,6 +1,29 @@
+    setClause.push(`isOpen='${isOpen}'`);
+  }
+
+  const query = `UPDATE milestones SET ${setClause.join(',')} WHERE id='${milestoneId}'`;
```
`query`는 `?`(placeholder)로 생성하고 `execute`에서 값들(`title`, `dueDate`, ...)을 배열과 함께 넣어주면 어떨까요? 저도 잘 설명하지 못할 것 같아서 자료를 좀 찾아봤는데 다음과 같은 장점이 있다고 합니다. 🙃

1. 성능 면에서 우수
위와 같은 방식을 prepared statement라고 부르는데 이 방식은 placeholder(`?`)가 있는 쿼리문과 값들이 따로 보내집니다. 먼저 쿼리문을 어떻게 수행할 것인지 준비되고 이후 값을 넣어서 쿼리의 결과를 반환합니다.
```js
connection.execute('select 1 + ? + ? as result', [5, 6], (err, rows) => {
  // rows: [ { result: 12 } ]
  // internally 'select 1 + ? + ? as result' is prepared first. On subsequent calls cached statement is re-used
});
```

```js
connection.prepare('select ? + ? as tests', (err, statement) => {
  // statement.parameters - array of column definitions, length === number of params, here 2
  // statement.columns - array of result column definitions. Can be empty if result schema is dynamic / not known
  // statement.id
  // statement.query

  statement.execute([1, 2], (err, rows, columns) => {
    // -> [ { tests: 3 } ]
  });

  // note that there is no callback here. There is no statement close ack at protocol level.
  statement.close();
});
```

`execute`를 placeholder를 활용해서 호출하면 쿼리문을 어떻게 실행할지 계산된 결과가 서버에서 캐시되기 때문에 성능면에서 더 낫다고 합니다. (다시 쿼리 실행 계획을 계산할 필요가 없으니까?)

2. SQL Injection 공격을 막을 수 있음

```js
const spoiled_data = '1; DROP TABLE users;'
const query        = `SELECT * FROM users where id=${spoiled_data}`;
```

이러한 코드는 다음과 같은 SQL문을 실행하게 됩니다. 😨

```sql
SELECT * FROM users where id=1; DROP TABLE users;
```

`execute`와 placeholder를 사용하면 위에 적은 것처럼 쿼리와 값이 따로 보내집니다. 먼저 쿼리의 실행 계획이 계산되고 필요한 값의 대체(`?`에 값을 넣는 일)가 수행되기 때문에 안전하다고 합니다.

cf.

`db.query('SQL... ? ... ? ...', [a, b])`와 같이 실행하면 클라이언트에서 `?`의 대체가 수행된다고 합니다. 이 역시 SQL Injection에 대해서 안전할 것 같기는 하지만 성능 상의 이점도 있으니 `execute`로 prepared statement를 활용하는 것이 좋을 것 같습니다.

> query() does parameter substitution on the client, and as a consequence we have a little bit more flexibility on what and how can be substituted.

훑어 본 자료들입니다.

1. https://www.npmjs.com/package/mysql2#using-prepared-statements
2. https://github.com/sidorares/node-mysql2/blob/master/documentation/Prepared-Statements.md
3. https://stackoverflow.com/questions/8263371/how-can-prepared-statements-protect-from-sql-injection-attacks
4. https://github.com/sidorares/node-mysql2/issues/553

thesulks commented on Backend/Routes/milestoneRoute.js:
```diff
@@ -44,4 +67,11 @@ router.get('/', async (req, res) => {
 });
 
+router.post('/', async (req, res) => {
+  const query = `INSERT INTO milestones(title, dueDate, description, isOpen) VALUES(?,?,?,?)`;
```
👍

marulloc committed:
[STYLE] prepared statement

[why]
for better query

[what]
alter standard statement to prepared statement

[etc]
etc...

thesulks reviewed (approved):
🤗👍

JeesooHaa reviewed (approved):
고생하셨습니다!
